network/endpoint_windows.go

```diff
@@ -14,16 +14,58 @@ import (
 	"github.com/Microsoft/hcsshim"
 )
 
+// ConstructEpName constructs endpoint name from netNsPath.
+func ConstructEpName(containerID string, netNsPath string, ifName string) string {
+	epName := ""
+	if netNsPath != "" {
+		splits := strings.Split(netNsPath, ":")
+		if len(splits) == 2 {
+			epName = splits[1]
+		} else {
+			epName = containerID
+		}
+		if len(epName) > 8 {
+			epName = epName[:8] + "-" + ifName
+			log.Printf("[net] constructed epName: %v", epName)
+		}
+	}
+	return epName
+}
+
 // newEndpointImpl creates a new endpoint in the network.
 func (nw *network) newEndpointImpl(epInfo *EndpointInfo) (*endpoint, error) {
+	// Ignore consecutive ADD calls for the same container.
+	if nw.Endpoints[epInfo.Id] != nil {
+		log.Printf("[net] Found existing endpoint %v, return immediately.", epInfo.Id)
+		return nw.Endpoints[epInfo.Id], nil
+	}
+
+	// Get Infrastructure containerID. Handle ADD calls for workload container.
+	epName := ConstructEpName(epInfo.ContainerID, epInfo.NetNsPath, epInfo.IfName)
+	log.Printf("[net] infraEpName: %v", epName)
+
+	hnsEndpoint, _ := hcsshim.GetHNSEndpointByName(epName)
+	if hnsEndpoint != nil {
+		log.Printf("[net] Found existing endpoint through hcsshim%v", epName)
+		log.Printf("[net] Attaching ep %v to container %v", hnsEndpoint.Id, epInfo.ContainerID)
+		if err := hcsshim.HotAttachEndpoint(epInfo.ContainerID, hnsEndpoint.Id); err != nil {
+			return nil, err
+		}
+		return nw.Endpoints[epName], nil
+	}
+
 	// Initialize HNS endpoint.
-	hnsEndpoint := &hcsshim.HNSEndpoint{
-		Name:           epInfo.Id,
+	hnsEndpoint = &hcsshim.HNSEndpoint{
+		Name:           epName,
 		VirtualNetwork: nw.HnsId,
 		DNSSuffix:      epInfo.DNS.Suffix,
 		DNSServerList:  strings.Join(epInfo.DNS.Servers, ","),
 	}
 
+	//enable outbound NAT
+	var enableOutBoundNat = json.RawMessage(`{"Type":  "OutBoundNAT"}`)
```

Comment:
Do we have unit test for testing this function? If its there, can we add one to test this property is set?

Reply:
I didn't write unit test for this. We can discuss this.